File: app/src/main/java/com/example/collections_and_maps/ui/benchmark/BenchmarkViewModel.java
package com.example.collections_and_maps.ui.benchmark;

import androidx.annotation.NonNull;
import androidx.core.util.Pair;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.example.collections_and_maps.R;
import com.example.collections_and_maps.models.benchmarks.Benchmark;
import com.example.collections_and_maps.models.benchmarks.ResultItem;

import java.util.ArrayList;
import java.util.List;

import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers;
import io.reactivex.rxjava3.core.Observable;
import io.reactivex.rxjava3.disposables.Disposable;
import io.reactivex.rxjava3.schedulers.Schedulers;

public class BenchmarkViewModel extends ViewModel {
    private final MutableLiveData<List<ResultItem>> itemsLiveData = new MutableLiveData<>();
    private final MutableLiveData<Integer> liveTextTV = new MutableLiveData<>();
    private final MutableLiveData<Integer> liveShowerMessages = new MutableLiveData<>();
    private final Benchmark benchmark;
    private Disposable disposable = Disposable.disposed();

    public BenchmarkViewModel(Benchmark benchmark) {
        this.benchmark = benchmark;
    }

    public LiveData<List<ResultItem>> getItemsLiveData() {
        return itemsLiveData;
    }

    public LiveData<Integer> getLiveTextTV() {
        return liveTextTV;
    }

    public LiveData<Integer> getLiveShowerMessages() {
        return liveShowerMessages;
    }

    public void onCreate() {
        itemsLiveData.setValue(benchmark.getItemsList(false));
    }

    public void startMeasure(@NonNull String inputtedValue) {
        if (disposable.isDisposed()) {
            final int value = checkValidateValue(inputtedValue);
            if (value < 0) {
                return;
            }

            liveTextTV.setValue(R.string.calcButtonStop);
            final List<ResultItem> items = benchmark.getItemsList(true);
            itemsLiveData.setValue(new ArrayList<>(items));

            disposable = Observable.fromIterable(items)
                    .filter(rItem -> !rItem.isHeader())
                    .flatMap(it -> Observable.fromCallable(() -> Pair.create(
                                    getIndex(it, items), getNewItem(it, value))
                    )).subscribeOn(Schedulers.computation())
                    .observeOn(AndroidSchedulers.mainThread())
                    .doFinally(() -> liveTextTV.setValue(R.string.calcButtonStart))
                    .subscribe(pair -> {
                          putIndex(pair, items);
                          putNewListItems(items);
                    }, Throwable::printStackTrace);
        } else {
            disposable.dispose();
        }
    }

    private int getIndex(ResultItem it, List<ResultItem> items) {
        return items.indexOf(it);
    }

    private ResultItem getNewItem(ResultItem it, int value) {
        return it.copy(it, benchmark.getMeasureTime(it, value));
    }

    private void putIndex(Pair <Integer, ResultItem> pair, List<ResultItem> items) {
        items.set(pair.first, pair.second);
    }

    private void putNewListItems(List <ResultItem> items) {
        final List<ResultItem> newItems = new ArrayList<>(items);
        itemsLiveData.setValue(newItems);
    }

    public int getSpan() {
        return benchmark.getSpan();
    }

    private int checkValidateValue(String inputtedValue) {
        int value = -1;
        Integer message = null;
        try {
            value = Integer.parseInt(inputtedValue);
        } catch (NumberFormatException e) {
            message = R.string.empty_input_value;
            e.printStackTrace();
        }
        liveShowerMessages.setValue(message);
        return value;
    }
}

File: app/src/main/java/com/example/collections_and_maps/ui/benchmark/BenchmarkViewModel.kt
package com.example.collections_and_maps.ui.benchmark

import androidx.core.util.Pair
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import com.example.collections_and_maps.R
import com.example.collections_and_maps.models.benchmarks.Benchmark
import com.example.collections_and_maps.models.benchmarks.ResultItem
import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers
import io.reactivex.rxjava3.core.Observable
import io.reactivex.rxjava3.disposables.Disposable
import io.reactivex.rxjava3.schedulers.Schedulers

class BenchmarkViewModel(private val benchmark: Benchmark) : ViewModel() {

    private val itemsLiveData = MutableLiveData<List<ResultItem>>()
    private val liveTextTV = MutableLiveData<Int>()
    private val liveShowerMessages = MutableLiveData<Int?>()
    private var disposable = Disposable.disposed()

    fun getItemsLiveData(): LiveData<List<ResultItem>> = itemsLiveData

    fun getLiveTextTV(): LiveData<Int> = liveTextTV

    fun getLiveShowerMessages(): LiveData<Int?> = liveShowerMessages

    fun onCreate() {
        itemsLiveData.value = benchmark.getItemsList(false)
    }

    fun startMeasure(inputtedValue: String) {
        if (disposable.isDisposed) {
            val value = checkValidateValue(inputtedValue)
            if (value < 0) {
                return
            }

            liveTextTV.postValue(R.string.calcButtonStop)
            val items = benchmark.getItemsList(true).toMutableList()
            itemsLiveData.postValue(ArrayList(items))

            disposable = Observable.fromIterable(items)
                .filter { rItem -> !rItem.isHeader() }
                .flatMap { it ->
                    Observable.fromCallable {
                        Pair.create(
                            items.indexOf(it),
                            it.copy(it, benchmark.getMeasureTime(it, value))
                        )
                    }
                }
                .subscribeOn(Schedulers.computation())
                .observeOn(AndroidSchedulers.mainThread())
                .doFinally { liveTextTV.postValue(R.string.calcButtonStart) }
                .subscribe({ pair ->
                    items[pair.first] = pair.second
                    itemsLiveData.postValue(ArrayList(items))
                }, { throwable -> throwable.printStackTrace() })
        } else {
            disposable.dispose()
        }
    }


    fun getSpan(): Int = benchmark.getSpan()

    private fun checkValidateValue(inputtedValue: String): Int {
        var value = -1
        var message: Int? = null
        try {
            value = inputtedValue.toInt()
        } catch (e: NumberFormatException) {
            message = R.string.empty_input_value
            e.printStackTrace()
        }
        return if (value == 0) {
            liveShowerMessages.value = R.string.OverZero
            -1
        } else {
            liveShowerMessages.value = message
            value
        }
    }
}
